File: architecture-skills/design-patterns/examples/composite/composite.swift
// Composite pattern in Swift.
//
// Files and directories conform to one Node protocol. The directory is
// a class because the tree is recursive, reference-shaped data; a
// closed set of node kinds could instead be an indirect enum.

protocol Node {
    var size: Int { get }
    func printTree(indent: Int)
}

struct File: Node {
    let name: String
    let size: Int

    func printTree(indent: Int) {
        print("\(String(repeating: " ", count: indent))\(name) (\(size) B)")
    }
}

final class Directory: Node {
    let name: String
    private var children: [Node] = []

    init(_ name: String) {
        self.name = name
    }

    @discardableResult
    func add(_ child: Node) -> Directory {
        children.append(child)
        return self
    }

    var size: Int {
        children.reduce(0) { $0 + $1.size }
    }

    func printTree(indent: Int) {
        print("\(String(repeating: " ", count: indent))\(name)/ (\(size) B)")
        for child in children {
            child.printTree(indent: indent + 2)
        }
    }
}

let project = Directory("project")
    .add(Directory("src")
        .add(File(name: "main.py", size: 1200))
        .add(File(name: "util.py", size: 450)))
    .add(Directory("docs").add(File(name: "readme.md", size: 800)))
    .add(File(name: ".gitignore", size: 20))

project.printTree(indent: 0)

File: architecture-skills/design-patterns/examples/composite/composite.rs
// Composite pattern in Rust.
//
// A trait plus Box<dyn Node> children gives the open-ended GoF form.
// When the set of node kinds is closed, a recursive enum
// (enum Node { File(..), Dir(..) }) is often the more idiomatic choice.

trait Node {
    fn size(&self) -> u64;
    fn print(&self, indent: usize);
}

struct File {
    name: String,
    size: u64,
}

impl File {
    fn new(name: &str, size: u64) -> Box<Self> {
        Box::new(Self { name: name.to_string(), size })
    }
}

impl Node for File {
    fn size(&self) -> u64 {
        self.size
    }
    fn print(&self, indent: usize) {
        println!("{:indent$}{} ({} B)", "", self.name, self.size);
    }
}

struct Directory {
    name: String,
    children: Vec<Box<dyn Node>>,
}

impl Directory {
    fn new(name: &str) -> Self {
        Self { name: name.to_string(), children: Vec::new() }
    }

    fn add(mut self, child: Box<dyn Node>) -> Self {
        self.children.push(child);
        self
    }
}

impl Node for Directory {
    fn size(&self) -> u64 {
        self.children.iter().map(|child| child.size()).sum()
    }
    fn print(&self, indent: usize) {
        println!("{:indent$}{}/ ({} B)", "", self.name, self.size());
        for child in &self.children {
            child.print(indent + 2);
        }
    }
}

fn main() {
    let project = Directory::new("project")
        .add(Box::new(
            Directory::new("src")
                .add(File::new("main.py", 1200))
                .add(File::new("util.py", 450)),
        ))
        .add(Box::new(Directory::new("docs").add(File::new("readme.md", 800))))
        .add(File::new(".gitignore", 20));

    project.print(0);
}
